src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -745,12 +745,23 @@ size_t ShenandoahHeap::initial_capacity() const {
   return _initial_size;
 }
 
-bool ShenandoahHeap::is_in(const void* p) const {
+bool ShenandoahHeap::is_in_bounds(const void* p) const {
   HeapWord* heap_base = (HeapWord*) base();
   HeapWord* last_region_end = heap_base + ShenandoahHeapRegion::region_size_words() * num_regions();
   return p >= heap_base && p < last_region_end;
 }
 
+bool ShenandoahHeap::is_in(const void* p) const {
+  if (is_in_bounds(p)) {
+    // Now check if region is in active state, unless we are moving
+    // objects during Full GC across the regions in not yet determinate state.
+    return is_full_gc_move_in_progress() ||
+           heap_region_containing(p)->is_active();
```

Comment:
Should this also check against the region bounds?

Reply:
Not sure I understand. The if-condition checks that we are pointing into heap. This means `heap_region_containing` always returns the region.

Reply:
heap_region_containing() returns the region for which bottom <= p < end, my question is if we should check if bottom <= p < top, in other words if p is within the region's currently allocated part (as opposed to the unallocated tail, if any).

Reply:
Right, we should check that as well. Done in new commit, but I have to re-run the tests now, because something might trigger on this.